File: javascript/tetris.js
const canvas = document.getElementById('tetrisCanvas');
const ctx = canvas.getContext('2d');
const nextCanvas = document.getElementById('nextCanvas');
const nextCtx = nextCanvas.getContext('2d');

const COLS = 10;
const ROWS = 20;
const CELL_SIZE = 30;

let field = Array(ROWS).fill().map(() => Array(COLS).fill(0));
let score = 0;
let activePiece = null;
let pieceX, pieceY;
let gameLoop = null;

const SHAPES = [
    [[1,1,1,1]],
    [[1,1],[1,1]],
    [[0,1,0],[1,1,1]],
    [[1,0,0],[1,1,1]],
    [[0,0,1],[1,1,1]],
    [[0,1,1],[1,1,0]],
    [[1,1,0],[0,1,1]]
];

function randomPiece() {
    const shape = SHAPES[Math.floor(Math.random() * SHAPES.length)];
    return shape.map(row => [...row]);
}

function spawnNewPiece() {
    activePiece = randomPiece();
    pieceX = Math.floor((COLS - activePiece[0].length) / 2);
    pieceY = 0;
    if (collision()) {
        gameOver();
    }
}

function collision() {
    for (let y = 0; y < activePiece.length; y++) {
        for (let x = 0; x < activePiece[y].length; x++) {
            if (activePiece[y][x] !== 0) {
                const fieldX = pieceX + x;
                const fieldY = pieceY + y;
                if (fieldX < 0 || fieldX >= COLS || fieldY >= ROWS || (fieldY >= 0 && field[fieldY][fieldX] !== 0)) {
                    return true;
                }
            }
        }
    }
    return false;
}

function merge() {
    for (let y = 0; y < activePiece.length; y++) {
        for (let x = 0; x < activePiece[y].length; x++) {
            if (activePiece[y][x] !== 0) {
                field[pieceY + y][pieceX + x] = 1;
            }
        }
    }
    clearLines();
    spawnNewPiece();
}

function clearLines() {
    let linesCleared = 0;
    for (let y = ROWS - 1; y >= 0; y--) {
        if (field[y].every(cell => cell !== 0)) {
            field.splice(y, 1);
            field.unshift(Array(COLS).fill(0));
            linesCleared++;
            y++;
        }
    }
    score += linesCleared * 100;
    document.getElementById('score').innerText = score;
}

function move(dx, dy) {
    pieceX += dx;
    pieceY += dy;
    if (collision()) {
        pieceX -= dx;
        pieceY -= dy;
        if (dy === 1) {
            merge();
        }
        return false;
    }
    return true;
}

function rotate() {
    const rotated = activePiece[0].map((_, idx) => activePiece.map(row => row[idx]).reverse());
    const oldPiece = activePiece;
    activePiece = rotated;
    if (collision()) {
        activePiece = oldPiece;
    }
}

function draw() {
    ctx.clearRect(0, 0, canvas.width, canvas.height);
    // поле
    for (let y = 0; y < ROWS; y++) {
        for (let x = 0; x < COLS; x++) {
            if (field[y][x]) {
                ctx.fillStyle = '#FF4B4B';
                ctx.fillRect(x * CELL_SIZE, y * CELL_SIZE, CELL_SIZE - 1, CELL_SIZE - 1);
            } else {
                ctx.strokeStyle = '#333';
                ctx.strokeRect(x * CELL_SIZE, y * CELL_SIZE, CELL_SIZE - 1, CELL_SIZE - 1);
            }
        }
    }
    // активная фигура
    for (let y = 0; y < activePiece.length; y++) {
        for (let x = 0; x < activePiece[y].length; x++) {
            if (activePiece[y][x]) {
                ctx.fillStyle = '#FF4B4B';
                ctx.fillRect((pieceX + x) * CELL_SIZE, (pieceY + y) * CELL_SIZE, CELL_SIZE - 1, CELL_SIZE - 1);
            }
        }
    }
}

function gameOver() {
    clearInterval(gameLoop);
    alert('Game Over! Score: ' + score);
}

function step() {
    move(0, 1);
    draw();
}

function init() {
    spawnNewPiece();
    draw();
    gameLoop = setInterval(step, 500);
    window.addEventListener('keydown', (e) => {
        switch(e.key) {
            case 'ArrowLeft': move(-1, 0); draw(); break;
            case 'ArrowRight': move(1, 0); draw(); break;
            case 'ArrowDown': move(0, 1); draw(); break;
            case 'ArrowUp': rotate(); draw(); break;
            case ' ': move(0, 1); while(move(0,1)) {}; draw(); break;
        }
    });
}

init();

File: cpp/tetris.cpp
#include <ncurses.h>
#include <cstdlib>
#include <ctime>
#include <vector>

const int WIDTH = 10;
const int HEIGHT = 20;

std::vector<std::vector<int>> field(HEIGHT, std::vector<int>(WIDTH, 0));
int score = 0;
bool gameOver = false;

std::vector<std::vector<std::vector<int>>> shapes = {
    {{1,1,1,1}},
    {{1,1},{1,1}},
    {{0,1,0},{1,1,1}},
    {{1,0,0},{1,1,1}},
    {{0,0,1},{1,1,1}},
    {{0,1,1},{1,1,0}},
    {{1,1,0},{0,1,1}}
};

std::vector<std::vector<int>> activePiece;
int pieceX, pieceY;

std::vector<std::vector<int>> getRandomPiece() {
    int idx = rand() % shapes.size();
    return shapes[idx];
}

bool collision() {
    for (size_t y = 0; y < activePiece.size(); y++) {
        for (size_t x = 0; x < activePiece[y].size(); x++) {
            if (activePiece[y][x]) {
                int fx = pieceX + x;
                int fy = pieceY + y;
                if (fx < 0 || fx >= WIDTH || fy >= HEIGHT || (fy >= 0 && field[fy][fx]))
                    return true;
            }
        }
    }
    return false;
}

void merge() {
    for (size_t y = 0; y < activePiece.size(); y++) {
        for (size_t x = 0; x < activePiece[y].size(); x++) {
            if (activePiece[y][x])
                field[pieceY + y][pieceX + x] = 1;
        }
    }
    // очистка линий
    for (int y = HEIGHT-1; y >= 0; ) {
        bool full = true;
        for (int x = 0; x < WIDTH; x++)
            if (!field[y][x]) { full = false; break; }
        if (full) {
            for (int i = y; i > 0; i--)
                field[i] = field[i-1];
            field[0] = std::vector<int>(WIDTH, 0);
            score += 100;
        } else y--;
    }
    // новая фигура
    activePiece = getRandomPiece();
    pieceX = WIDTH/2 - activePiece[0].size()/2;
    pieceY = 0;
    if (collision()) gameOver = true;
}

void move(int dx, int dy) {
    pieceX += dx;
    pieceY += dy;
    if (collision()) {
        pieceX -= dx;
        pieceY -= dy;
        if (dy == 1) merge();
    }
}

void rotate() {
    std::vector<std::vector<int>> rotated(activePiece[0].size(), std::vector<int>(activePiece.size(), 0));
    for (size_t i = 0; i < activePiece.size(); i++)
        for (size_t j = 0; j < activePiece[i].size(); j++)
            rotated[j][activePiece.size()-1-i] = activePiece[i][j];
    auto old = activePiece;
    activePiece = rotated;
    if (collision()) activePiece = old;
}

void draw() {
    clear();
    printw("Score: %d\n", score);
    for (int y = 0; y < HEIGHT; y++) {
        for (int x = 0; x < WIDTH; x++) {
            bool cell = field[y][x];
            if (!cell && pieceY <= y && y < pieceY + (int)activePiece.size() &&
                pieceX <= x && x < pieceX + (int)activePiece[0].size() &&
                activePiece[y-pieceY][x-pieceX])
                cell = true;
            printw(cell ? "[]" : "  ");
        }
        printw("\n");
    }
    refresh();
}

int main() {
    srand(time(0));
    initscr();
    cbreak();
    noecho();
    keypad(stdscr, TRUE);
    nodelay(stdscr, TRUE);
    curs_set(0);

    activePiece = getRandomPiece();
    pieceX = WIDTH/2 - activePiece[0].size()/2;
    pieceY = 0;

    while (!gameOver) {
        draw();
        int ch = getch();
        switch(ch) {
            case KEY_LEFT:  move(-1,0); break;
            case KEY_RIGHT: move(1,0); break;
            case KEY_DOWN:  move(0,1); break;
            case KEY_UP:    rotate(); break;
            case 'q': gameOver = true; break;
        }
        move(0,1);
        napms(300);
    }
    clear();
    printw("GAME OVER! Score: %d\n", score);
    refresh();
    napms(3000);
    endwin();
    return 0;
}
